cypress-io/cypress
# fix: support ESM projects using TypeScript with ts-node/esm

- Closes https://github.com/cypress-io/cypress/issues/22096
- Closes https://github.com/cypress-io/cypress/issues/21939

### User facing changelog


Support TypeScript projects using `"type": "module"` using `ts-node/esm`.

### Additional details


The problem is if you are using `"type": "module"`, Node.js will use it's ESM loader to load your code. If you are using a `ts` extension, which Node.js does not know about out of the box, you get the various errors we've seen reported.

The original solution was to use `esbuild` to try load TS files if the initial `require('cypress.config...`) failed, but it's not working as expected, as [explained here](https://github.com/cypress-io/cypress/issues/22074#issuecomment-1147097883). I removed the `esbuild` code path for now. I noted this [here](https://github.com/cypress-io/cypress/issues/22074) so we can bring it back in the future, if needed.

The solution introduced here is to use the [ESM support that `ts-node` ships](https://github.com/TypeStrong/ts-node/issues/1007), as recommended in #21939. We rely on `ts-node` in the project and already ship it, so it makes sense to continue to use it where we can.

### Steps to test


0. Reproduce https://github.com/cypress-io/cypress/issues/22096/ with the latest binary from npm and verify problem
1. Grab this branch and build the binary locally
2. Open the repro with the new binary, it should work now


### How has the user experience changed?


Can correctly load `cypress.config.ts` for `"type": "module"` projects.

## The workflow is now:

<img width="915" alt="image" src="https://user-images.githubusercontent.com/19196536/172330104-ddc18836-e293-472b-a412-379f2e288581.png">

## The old workflow (shows problem):

<details>
<summary>Show old workflow diagram</summary>
<img width="828" alt="image" src="https://user-images.githubusercontent.com/19196536/172334805-54ed333c-087e-423c-b0ef-bef3afb860e5.png">
</details>


### PR Tasks
